index.bs

```diff
@@ -52,10 +66,169 @@ dictionary StorageBucketOptions {
+
+1. If |quota| is less than or equal to zero, then return failure.
+
+1. Let |persisted| be false.
+
+1. If |options|["{{StorageBucketOptions/persisted}}"] is true, then:
+
+   1. Let |permission| be the result of [=/requesting permission to use=] "<code>persistent-storage</code>".
+
+   1. If |permission| is "{{PermissionState/granted}}", then set |persisted| to true.
+
+1. Let |bucket| be the [=/storage bucket=] named |name| in |storageKey| or null otherwise.
+
+1. If |bucket| is non-null and |bucket|'s [=bucket expiration=] is less than or equal to now, then:
+
+   1. Remove |bucket|.
+
+   1. Set |bucket| to null.
+
+1. If |bucket| is null, then:
+
+   1. Let |bucket| be a new [=/storage bucket=] with name |name|
+
+   1. Set |bucket|'s [=bucket durability|durability=] to |options|["{{StorageBucketOptions/durability}}"] if it exists.
+
+   1. Set |bucket|'s [=bucket quota|quota=] to |quota|.
+
+1. If |persisted| is true, set |bucket|'s [=/bucket mode=] to "<code>persistent</code>".
+
+1. Set |bucket|'s [=bucket expiration|expiration=] to |expires|.
+
+1. Let |storageBucket| be a new {{StorageBucket}}.
+
+1. Set |storageBucket|'s [=/storage bucket=] to |bucket|.
+
+1. Return |storageBucket|.
+
+</div>
+
+<div algorithm>
+
+To <dfn>validate a bucket name</dfn> given string |name|, run the following steps:
+
+1. If |name| contains any character that is not [=ASCII lower alpha=], [=ASCII digit=], U+005F (_), or U+002D(-), then return failure.
+
+1. If |name| [=string/length=] is 0 or exceeds 64, then return failure.
+
+1. If |name| begins with U+005F (_) or U+002D(-), then return failure.
+
+1. Return.
+
+</div>
+
+When not null, {{StorageBucketOptions/durability}} is a hint to the user agent
+specifying the desired {{StorageBucket/durability}}. The user agent MAY
+create a new {{StorageBucket}} with this [=bucket durability=]. The user agent
+MUST NOT modify the [=bucket durability=] of an existing bucket.
+
 <h3 id="storage-bucket-delete">Deleting a bucket</h3>
+
+<div algorithm>
+
+The <dfn method for="StorageBucketManager">delete(|name|)</dfn> method steps are:
+
+1. Let |environment| be [=/this=]'s [=/relevant settings object=].
+
+1. Let |storageKey| be the result of running [=obtain a storage key=] given |environment|.
+
+1. If |storageKey| is failure, then [=exception/throw=] a "{{SecurityError}}" {{DOMException}} and abort these steps.
+
+1. Let |p| be [=a new promise=].
+
+1. Run the following steps [=in parallel=]:
+
+    1. If the result of [=validate a bucket name=] with |name| is failure, then [=/reject=] |p| with an {{InvalidCharacterError}} and abort these steps.
+
+    1. Let |bucket| be the [=/storage bucket=] named |name| in |storageKey| if one exists. Otherwise return.
+
+    1. Remove |bucket|.
+
+    1. [=/Resolve=] |p|.
+
+1. Return |p|.
+
+</div>
+
+Issue: [[Storage]] needs to define bucket removal. The bucket needs to be internally marked as
```

Comment:
Is this more than removing the bucket from the bucket map?  Or do we need to hold on to it somewhere until the last window with an open reference goes away?  Can we monkey patch it?

Reply:
We don't expect to hold on to it until the last window goes away. So we'd like to define how this would behave in the Storage spec when a bucket is removed but there is still a reference. Probably will need to update text under Management here as well. https://storage.spec.whatwg.org/#management

Do you know of an example of a monkey patch I can follow for format?

Reply:
I'm wondering, do you really need to monkey patch the storage spec here?  Or can you just remove the entry from the bucket map?  What other steps do you need to take?

I'll cc you on a separate conversation with examples of monkey patching.